README.md
# deno_mongo_schema

Extend from the Node.js
**[mongodb client v5.3.0](https://github.com/mongodb/node-mongodb-native)**,
support Schema and extend some API.

[![ci](https://github.com/jiawei397/deno_mongo_schema/actions/workflows/ci.yml/badge.svg)](https://github.com/jiawei397/deno_mongo_schema/actions/workflows/ci.yml)
[![tag](https://img.shields.io/badge/deno-v1.32.5-green.svg)](https://github.com/denoland/deno)

> Breaking changes on v1.0.0.
>
> I have to switch client from [deno_mongo](https://deno.land/x/mongo) to
> Node.js. Because it is officially maintained, it looks more robust.
>
> If it is to be used in production, reconnecting after wire breakage is an
> important function.

## hooks

```ts
import {
  BaseSchema,
  getDB,
  getModel,
  MongoFactory,
  MongoHookMethod,
  Prop,
  Schema,
  UpdateExOptions,
} from "https://deno.land/x/deno_mongo_schema@v1.0.2/mod.ts";
import type { Document } from "https://deno.land/x/deno_mongo_schema@v1.0.2/mod.ts";

await MongoFactory.forRoot("mongodb://localhost:27017/test");

@Schema()
class User extends BaseSchema {
  @Prop()
  age!: number;

  @Prop({
    required: true,
  })
  name!: string;

  @Prop({
    default: Date.now,
    expires: 60, // seconds
  })
  expires?: Date;
}

const UserSchema = SchemaFactory.createForClass(User);

UserSchema.pre(
  MongoHookMethod.update,
  function (filter: Document, doc: Document, options?: UpdateExOptions) {
    console.log("----pre----", filter, doc, options);
    if (!doc.$set) {
      doc.$set = {};
    }
    doc.$set.modifyTime = new Date();
  },
);

UserSchema.post(MongoHookMethod.findOneAndUpdate, function (doc) {
  console.log("----post----", doc);
  doc.name = "haha";
});

const userModel = await MongoFactory.getModel(User);

const id = await userModel.insertOne({
  "name": "zhangsan",
  "age": 18,
});

UserSchema.post(MongoHookMethod.findOne, function (doc) {
  console.log("----post---findOne----", doc);
});

// console.log(id);
const info = await userModel.findById(id, {
  projection: {
    name: 1,
  },
});
console.log(info);

UserSchema.post(MongoHookMethod.findMany, function (doc) {
  console.log("----post---findMany----", doc);
});

const arr = await userModel.findMany({});
console.log(arr);

UserSchema.post(MongoHookMethod.delete, function (doc) {
  console.log("----post---delete----", doc);
});

const del = await userModel.deleteOne({
  name: "zhangsan",
});
console.log(del);

const delMulti = await userModel.deleteMany({
  name: "zhangsan",
});
console.log(delMulti);
```

Here are some useful APIs:

- find
  - findById
  - findOne
  - findMany
  - countDocuments
  - aggregate
  - distinct
- insert
  - insertOne
  - insertMany
- update
  - findByIdAndUpdate
  - findOneAndUpdate
  - updateMany
  - updateOne
- delete
  - deleteMany
  - deleteOne/findOneAndDelete
  - deleteById/findByIdAndDelete
- index
  - syncIndexes
  - dropIndexes
  - listIndexes
  - createIndexes
- drop collection
  - drop

## Virtual

Or you can use virtual like this:

```ts
import {
  BaseSchema,
  MongoFactory,
  Prop,
  Schema,
} from "https://deno.land/x/deno_mongo_schema@v1.0.2/mod.ts";

await MongoFactory.forRoot("mongodb://localhost:27017/test");

@Schema()
class User extends BaseSchema {
  @Prop()
  group!: string;

  @Prop()
  title!: string;
}

@Schema()
class Role extends BaseSchema {
  @Prop()
  userId!: string;

  @Prop()
  name!: string;
}

const RoleSchema = SchemaFactory.createForClass(Role);

RoleSchema.virtual("user", {
  ref: User,
  localField: "userId",
  foreignField: "_id",
  justOne: true,
  isTransformLocalFieldToObjectID: true,
});

// Role.populate("user", {
//   // _id: 0,
//   group: 1,
//   // title: 1,
// });
// Role.populate("user", "group");
// Role.populate("user", "-group -createTime");
// Role.populate("user", "title group");

const roleModel = await MongoFactory.getModel(Role);

// roleModel.insertOne({
//   userId: id,
//   name: "normal",
// });

console.log(
  await roleModel.findMany({}, {
    projection: {
      name: 1,
      userId: 1,
    },
    // skip: 1,
    // limit: 1,
    populates: {
      // user: {
      //   // _id: 0,
      //   group: 1,
      //   title: 1,
      // },
      // user: "group",
      user: true,
      // user: "-_id -title",
    },
  }),
);
```

## Special collection name

If you donnot want to use the default collection name, you must regiter it by
yourself.

```ts
SchemaFactory.createForClass(Role, "mongo_test_schema_roles");
```

Then if you still want to use virtual, you must use your registered name instead
of Schema Class.

```ts
UserSchema.virtual("role", {
  ref: "mongo_test_schema_roles",
  localField: "roleId",
  foreignField: "_id",
  justOne: true,
  // isTransformLocalFieldToObjectID: true,
  // isTransformObjectIDToLocalField: true
});
```

## TODO

- [x] Modify schema as a decorator
- [x] Unit
- [x] Configurable whether to convert _id
- [x] Configurable the createTime and modifyTime
- [x] Switch the underlying layer to the official Node.js version library


CHANGELOG.md
# [](https://github.com/jiawei397/deno_mongo_schema/compare/v0.11.1...v) (2023-04-27)


### Bug Fixes

* no bson now ([a8ec988](https://github.com/jiawei397/deno_mongo_schema/commit/a8ec98876c04f2af0558d107abbc7315b37c7266))


### Features

* **cache:** clear timeout when close ([e1b4fe2](https://github.com/jiawei397/deno_mongo_schema/commit/e1b4fe2593c3e75cdc038631405e9228a5aa3fea))
* **model:** filter type use any current ([b3a3ffc](https://github.com/jiawei397/deno_mongo_schema/commit/b3a3ffc617081e78e9da9326993a1b71694453bd))
* update mongodb client to 5.3.0 ([f28a398](https://github.com/jiawei397/deno_mongo_schema/commit/f28a3989f7acb90d02ce9aeb03998dd9074d8797))
* use npm:mongodb ([454b36f](https://github.com/jiawei397/deno_mongo_schema/commit/454b36f96a7ee752b44927d8e9c2d5070dc7425b))



## [0.11.1](https://github.com/jiawei397/deno_mongo_schema/compare/v0.11.0...v0.11.1) (2023-03-30)


### Features

* **type:** RequiredId not extends BaseSchema ([d52d67d](https://github.com/jiawei397/deno_mongo_schema/commit/d52d67df5ab3ab023c7a41794bb0b30e93f0b8d5))



# [0.11.0](https://github.com/jiawei397/deno_mongo_schema/compare/v0.10.5...v0.11.0) (2023-03-30)


### Features

* **model:** inject indexes ([53b44f7](https://github.com/jiawei397/deno_mongo_schema/commit/53b44f70dc26a5d5b99ea35a1b18fdc09c74cb20))
* **type:** find only set id and _id be required ([c338aa3](https://github.com/jiawei397/deno_mongo_schema/commit/c338aa3ef021df9907e168f7c402ac42cafd33dd))
* **types:** add Simplify ([6aa654f](https://github.com/jiawei397/deno_mongo_schema/commit/6aa654ff7ece839adbe18fccc349ec61991a5b2b))



## [0.10.5](https://github.com/jiawei397/deno_mongo_schema/compare/v0.10.4...v0.10.5) (2022-06-24)



## [0.10.4](https://github.com/jiawei397/deno_mongo_schema/compare/v0.10.3...v0.10.4) (2022-06-24)


### Features

* **reflect:** update reflect ([25fba22](https://github.com/jiawei397/deno_mongo_schema/commit/25fba22249ea2c6a1ae7375318acf4da92d30c8f))



## [0.10.3](https://github.com/jiawei397/deno_mongo_schema/compare/v0.10.2...v0.10.3) (2022-06-22)


### Features

* **model:** find return null instead of undefined ([f020126](https://github.com/jiawei397/deno_mongo_schema/commit/f0201269c80f46f35a4d803ac67d3f5544bd9483))



## [0.10.2](https://github.com/jiawei397/deno_mongo_schema/compare/v0.10.1...v0.10.2) (2022-06-22)


### Features

* **model:** find type required ([c2c570f](https://github.com/jiawei397/deno_mongo_schema/commit/c2c570f243a83b07a1e9e9d6ac31a82822bd0928))



## [0.10.1](https://github.com/jiawei397/deno_mongo_schema/compare/v0.10.0...v0.10.1) (2022-06-22)



# [0.10.0](https://github.com/jiawei397/deno_mongo_schema/compare/v0.9.3...v0.10.0) (2022-06-22)


### Features

* **model:** only find return id ([625d35a](https://github.com/jiawei397/deno_mongo_schema/commit/625d35abe6290cec2938d77b81fe3e34f070e572))



## [0.9.3](https://github.com/jiawei397/deno_mongo_schema/compare/v0.9.2...v0.9.3) (2022-05-05)


### Bug Fixes

* **index:** support expires alias ([5ed916f](https://github.com/jiawei397/deno_mongo_schema/commit/5ed916f20a16456d0181863e1596128a26fe3822))



## [0.9.2](https://github.com/jiawei397/deno_mongo_schema/compare/v0.9.1...v0.9.2) (2022-04-25)



## [0.9.1](https://github.com/jiawei397/deno_mongo_schema/compare/v0.9.0...v0.9.1) (2022-03-25)



# [0.9.0](https://github.com/jiawei397/deno_mongo_schema/compare/v0.8.5...v0.9.0) (2022-03-25)


### Performance Improvements

* change SchemaDecorator and Schema ([e83f6f8](https://github.com/jiawei397/deno_mongo_schema/commit/e83f6f8ae6ad27b2d2ecd01837a162e8befd02c5))



## [0.8.5](https://github.com/jiawei397/deno_mongo_schema/compare/v0.8.4...v0.8.5) (2022-03-25)


### Features

* **schema:** can define createTime and modifyTime ([e39f9fb](https://github.com/jiawei397/deno_mongo_schema/commit/e39f9fbe88451e3ca2a754ff56d35146eb736cde))


### Performance Improvements

* **model:** use #schema directly ([0cb6498](https://github.com/jiawei397/deno_mongo_schema/commit/0cb64980f402862c72ef12661488c34a82ee532d))



## [0.8.4](https://github.com/jiawei397/deno_mongo_schema/compare/v0.8.3...v0.8.4) (2022-03-25)


### Bug Fixes

* **model:** findOneAndUpdate can insert createTime ([f06ff24](https://github.com/jiawei397/deno_mongo_schema/commit/f06ff245704257d46901867b95bed9570be0197a))



## [0.8.3](https://github.com/jiawei397/deno_mongo_schema/compare/v0.8.2...v0.8.3) (2022-03-12)



## [0.8.2](https://github.com/jiawei397/deno_mongo_schema/compare/v0.8.1...v0.8.2) (2022-03-02)


### Bug Fixes

* **model:** transferid  conside doc maybe null ([a4b11dc](https://github.com/jiawei397/deno_mongo_schema/commit/a4b11dce2ba20e3d98739dcdbeffa09a78419a7b))



## [0.8.1](https://github.com/jiawei397/deno_mongo_schema/compare/v0.8.0...v0.8.1) (2022-02-24)


### Features

* update deno_mongo to 0.29.2 ([24bd732](https://github.com/jiawei397/deno_mongo_schema/commit/24bd732befb9f3014a1c2b26e113973f5abdd370))



# [0.8.0](https://github.com/jiawei397/deno_mongo_schema/compare/v0.7.5...v0.8.0) (2022-02-24)


### Features

* **model:** populate support array ([186bbcb](https://github.com/jiawei397/deno_mongo_schema/commit/186bbcbc12d69c424d8c762580a4c7e9190b9e85))



## [0.7.5](https://github.com/jiawei397/deno_mongo_schema/compare/v0.7.4...v0.7.5) (2022-02-22)


### Bug Fixes

* **model:** find options ([e274b32](https://github.com/jiawei397/deno_mongo_schema/commit/e274b324e35167c428a9edcdfb2695d0ceb0acd0))



## [0.7.4](https://github.com/jiawei397/deno_mongo_schema/compare/v0.7.3...v0.7.4) (2022-02-22)


### Bug Fixes

* **model:** this ([b202ecb](https://github.com/jiawei397/deno_mongo_schema/commit/b202ecb48004e545a717661c0891396ee00dc5b9))



## [0.7.3](https://github.com/jiawei397/deno_mongo_schema/compare/v0.7.2...v0.7.3) (2022-02-22)


### Bug Fixes

* **model:** format bsonId deal $or and $and ([7081260](https://github.com/jiawei397/deno_mongo_schema/commit/7081260f513c7be8b996509593210585c524cce4))




example/hooks.ts
// deno-lint-ignore-file no-explicit-any
import {
  BaseSchema,
  MongoFactory,
  MongoHookMethod,
  Prop,
  Schema,
  SchemaFactory,
  UpdateExOptions,
} from "../mod.ts";
import type { Document } from "../mod.ts";
import { dbUrl } from "../tests/common.ts";

await MongoFactory.forRoot(dbUrl);

@Schema()
class User extends BaseSchema {
  @Prop()
  age!: number;

  @Prop({
    required: true,
    index: true,
    // index: "text",
  })
  name!: string;

  @Prop({
    default: Date.now,
    expires: 60, // seconds
  })
  expires?: Date;
}

const UserSchema = SchemaFactory.createForClass(User);

UserSchema.pre(
  MongoHookMethod.update,
  function (filter: Document, doc: Document, options?: UpdateExOptions) {
    console.log("----pre----", filter, doc, options);
    if (!doc.$set) {
      doc.$set = {};
    }
    doc.$set.modifyTime = new Date();
  },
);

UserSchema.post(MongoHookMethod.findOneAndUpdate, function (doc: any) {
  console.log("----post----", doc);
  doc.name = "haha";
});

const model = await MongoFactory.getModel(User);

await model.deleteMany({});

const id = await model.insertOne({
  "name": "zhangsan",
  "age": 18,
});

// const wangwuInfo = await model.save({
//   "name": "wangwu",
//   "age": 20,
// });
// console.log("wangwuInfo", wangwuInfo);

UserSchema.post(MongoHookMethod.findOne, function (doc: any) {
  console.log("----post---findOne----", doc);
});

// console.log(id);
const info = await model.findById(id, {
  projection: {
    name: 1,
  },
});
console.log("info: ", info);

UserSchema.post(MongoHookMethod.findMany, function (doc: any) {
  console.log("----post---findMany----", doc);
});

const arr = await model.findMany({});
console.log("findMany: ", arr);

UserSchema.post(MongoHookMethod.delete, function (doc: any) {
  console.log("----post---delete----", doc);
});

// const del = await model.deleteOne({
//   name: "zhangsan",
// });
// console.log(del);

// const delMulti = await model.deleteMany({
//   name: "zhangsan",
// });
// console.log(delMulti);

// const res = await model.findByIdAndUpdate(id, {
//   $set: {
//     name: "lisi",
//   },
// });
// console.log("lisi", res);

const res2 = await model.findByIdAndUpdate(id, {
  $set: {
    name: "lisi2",
  },
}, {
  new: true,
});
console.log("lisi2", res2?.id, res2?.name);

const res3 = await model.findOneAndUpdate({
  name: "lisi2",
}, {
  $set: {
    name: "wangwu",
  },
}, {
  new: true,
});
console.log("res3", res3);
